### PR TITLE
meson: run_command, check : false

### PR DESCRIPTION
**Pull request by johandc**
The default for run_command in meson has been to ignore the result. This setting will change to check = true in the future. Because picolibc relies on the default option being = false, explicitly add check = false to kwargs of all run_command calls.

Building without an explicit check kwarg generates a warning on latest meson.

### CODE REVIEW

**Comment by rdiez** [comment]
Ignoring the exit code of external commands is usually asking for trouble. That is probably why Meson is changing the default. Lack of checking is normally an indication of deficient error handling.

Why are you saying that "Because picolibc relies on the default option being = false"?

From your  patch, say for example that "gcc -print-search-dirs" fails to run. Shouldn't the Picolibc build system stop straight away then?

**Comment by johandc** [comment]
> Why are you saying that "Because picolibc relies on the default option being = false"?

Because the current default value is false. And changing to true, would change the behavior. The purpose of this commit is not to change the behavior, but to quench the warning about the default changing to true. 

I think it should be up to Keith to decide, whether any of these should be changed into true. If they can all be set to true, it would be nice to set them now, because the deprecation period is set to be very long for this one. And avoiding the warning in the mean-time would be nice.

**Comment by rdiez** [comment]
Explicitly changing "check" to false just to prevent a new Meson warning helps hide what seems to be a code defect in Picolibc's build system.

It is effectively subverting the value of the new Meson warning, which has made us realise of this shortcoming in the first place.

**Comment by johandc** [comment]
I'm not sure Keith would like to build to fail if those run_command fails. I suppose it's written as a sort of configure check, and it's designed to fail sometimes when commands are not available. So just setting them all to true, or waiting for the default value to change, is presumably not the correct solution. Better look into each individual usage and make that decision.

**Comment by keith-packard** [comment]
Oh, this is cool. I had no idea meson was ignoring the result from these commands. In any case, I think you're both on solid ground here. Because meson is changing the default for this option (a questionable plan, imho), we need to set the value explicitly everywhere to avoid surprises when the default does change. What the right value is presumably depends upon whether we need to allow failures or not. How we get from the current state (relying on the default) to the desired state (explicitly stating whether we want to ignore or not) is the question here, and I'm happy to leave that process up to you; either setting every one of them to 'false' and then slowly changing some to 'true', or setting them all to the desired value immediately, or some mix of the two.

**Comment by yashi** [comment]
Checking all eight occurrences of `run_command()` in `meson.build` shows us that

- 4 are calling `cc` with various options, `-dumpmachine`, `-print-sysroot`, `-print-search-dirs`, and `--print-multi-lib'`, without any other input.
- 2 are to get the real path for the given path with `realpath(1)`.
- 1 is `expr`, which we should replace with Meson's `.split()`, to get an install dir from compiler's spec.
- 1 is `sed` to extract encoding names from `newlib/libc/iconv/encoding.aliases`.

At all call sites, the given _commands_ can't return non zero for meson to continue.  we expect all commands to return something via stdout.  So, I vote for `check : true` for all occurrences.

**Comment by johandc** [comment]
I agree, i have updated the PR to change to true.

File: meson.build [hunk 1]
```diff
@@ -408,9 +408,9 @@ endif
 specs_dir_option = get_option('specsdir')
 if specs_dir_option == ''
   search_cmds = cc.cmd_array() + ['-print-search-dirs']
-  install_dir=run_command(search_cmds).stdout().split('\n')[0]
+  install_dir=run_command(search_cmds, check : true).stdout().split('\n')[0]
   # Meson 0.56 adds a 'substring' method which can be used here
-  specs_dir=run_command(['expr', install_dir, ':', 'install: *\(.*\)']).stdout().split('\n')[0]
+  specs_dir=run_command(['expr', install_dir, ':', 'install: *\(.*\)'], check : true).stdout().split('\n')[0]
```
Comment (keith-packard) [inline meson.build:413]: Looks like this one needs to be 'false' as expr will exit with non-zero status when the result is empty, and that is checked below.

**Review by keith-packard — CHANGES_REQUESTED** [review]
This looks super cool; thanks for getting this ready before meson switches the default.

File: meson.build (hunk 1 above)
Comment (rdiez) [reply meson.build:413]: The right thing to do would be to check the exact status code: "Exit status is 0 if EXPRESSION is neither null nor 0, 1 if EXPRESSION is null or 0, 2 if EXPRESSION is syntactically invalid, and 3 if an error occurred.". So any other exit code should be translated to an error. I hope that Meson can do that.

Comment (keith-packard) [reply meson.build:413]: Yup, looks like you can get the exit status by stashing the run_command result in a variable and looking at the .returncode value. https://mesonbuild.com/External-commands.html

**Comment by eli-schwartz** [comment]
> Because meson is changing the default for this option (a questionable plan, imho),

For the record, when Meson says "this default will change in the future" what we actually mean is "maybe in a few years from now, maybe for Meson 1.0, who knows".

There is no chance of us flipping the default without a very long deprecation indeed.

**Comment by keith-packard** [comment]
This looks good; I've created a rebased branch and am running CI on that before merging this PR.

**Comment by keith-packard** [comment]
All merged in via the command line. Thanks for your contribution!

**Pull request closed without merging**